Bug 1453190 - Fix pod update operation

pravisankar opened the pull request:
Use pod sandbox ID to update the pod as opposed to pod container ID.
OVS flow note identified by sandbox ID is desired as network namespace
is held by the pod sandbox and pod could have many containers and single
container ID may not represent all the pod ovs flows.

Since we can't use kubelet 'Host' (explanation refer commit: f1118459),
we use runtime shim endpoint to connect to runtime service using gRPC.
This is the same mechanism used by kubelet(GenericKubeletRuntimeManager)
to talk to runtime service(docker/rkt).

bug 1453190

pravisankar commented:
@openshift/networking @dcbw PTAL

pravisankar commented:
[test]

danwinship commented on pkg/sdn/plugin/pod_linux.go:
```diff
@@ -228,6 +234,34 @@ func podIsExited(p *kcontainer.Pod) bool {
+	}
+
+	// Kubelet starts asynchronously and when we get an Update op, kubelet may not have created runtime endpoint.
+	// So try couple of times before bailing out (~2 second timeout).
```
2 seconds seems potentially like not enough...

danwinship commented:
What bug exactly does this fix?

(Code basically LGTM but this is @dcbw's area of expertise.)

danwinship commented:
So this only affects the re-VNID-ing case, not the resync-pods-at-startup case? In that case we could just change the way multitenant.go:updatePodNetwork() works, so that rather than going pod-by-pod, it just operates on the DumpFlows output as a whole, changing the VNID of every affected pod. That might end up being simpler?

dcbw commented:
@pravisankar I like the solution of talking to the runtime directly, I should have done that for 14665 too.  For this bug though, could we put the runtime related stuff into update.go once 14665 gets merged, rather than in pod_linux.go?

I think pod_linux.go and pod.go should just consume the given sandbox ID which would get passed from node.go (or update.go) instead.  Since Update is the only operation that requires talking to the runtime service, might make sense to put all that into update.go.  Other than that, looks generally good to me!

pravisankar commented:
@dcbw pod.go/pod_linux.go consuming sandbox ID seems good segregation, Let me rebase my changes on top of https://github.com/openshift/origin/pull/14665 and will move relevant changes to node/update.go

pravisankar commented:
@danwinship we can not directly work on the dump flows output in multitenant.go:updatePodNetwork(). We have to go through pod by pod (CNI_UPDATE per pod) so that CNI pod manager updates its bookkeeping (runningPods field, used for hostport/multicast handling). I think we can get dumpFlows() output once and reuse it for all the pods, this needs some fiddling with CNI request/response. I'll work on this change.

danwinship commented:
@pravisankar well #14893 depends on the changes here to use the runtime anyway, right? So in that case it makes sense to keep this like it is for now, although I'd really like to simplify the VNID-changing code at some point in the future

pravisankar commented:
@dcbw @danwinship Updated, ptal

knobunc commented:
[test]

knobunc commented:
[test] last flaked on https://github.com/openshift/origin/issues/14043

knobunc commented:
[test] [merge]

smarterclayton commented:
Hold on - so now our networking code depends on calling CRI in order to work correctly?

smarterclayton commented:
Removing merge tag temporarily.

Why is it good for SDN to make calls to the CRI directly?

dcbw commented:
> Why is it good for SDN to make calls to the CRI directly?

@smarterclayton because Kubernetes doesn't support changes to pods while they are running, but we do for things like merge/split tenant networks.  And that requires that we get sandbox IDs from the runtime, because we need to manipulate the sandbox to do those updates.  We used to get this information from kubelet directly via the network plugin interface, but we no longer can do that with the move to remote runtimes.  That's actually a good thing; it puts us on the same footing as all other CNI plugins.

Furthermore, I'm going to port https://github.com/openshift/origin/pull/14665 over to use these runtime interfaces since that better than banging on Docker directly and will actually work with CRI-O too.

smarterclayton commented:
Does the kubelet expect other daemons to be making calls to the CRI?  @derekwaynecarr @sjenning

sjenning commented:
@smarterclayton i'm not sure of any hard requirement that the kubelet have exclusive access to the CRI.  We have PLEG which should the kubelet if anything it cares about changes.

Am I correct in understanding that the only CRI that is being made in this change is querying the sandbox ID?

dcbw commented:
@smarterclayton I have to imagine CRI runtimes are safe here, since kubelet calls into the the runtimes from a couple different goroutines.  As to whether kubelet handles changes made underneath it, we're essentially doing "docker stop".  I realize it's more complicated than that, but looking at the dockershim code it looks like ListPodSandbox() doesn't affect any global state, and while it does read checkpoints, the checkpoint store is required to be thread-safe.

dcbw commented:
> Am I correct in understanding that the only CRI request that is being made in this change is querying the sandbox ID?

@sjenning yes, but I want to do a follow-up that will call StopPodSandbox() to replace the slighly horrible docker equivalent that #14665 currently does.

eparis commented:
[test]

pravisankar commented:
Jenkins failed to highlight actual error:
pkg/sdn/plugin/pod_test.go:332: unknown cniserver.PodRequest field 'ContainerId' in struct literal
pkg/sdn/plugin/pod_test.go:427: unknown cniserver.PodRequest field 'ContainerId' in struct literal
Fixed and Updated, please [test] again.

knobunc commented:
@sjenning @smarterclayton @derekwaynecarr are your concerns addressed?  Can we merge this?

Thanks

sjenning commented:
@knobunc @dcbw i guess i'm still not complete clear on what a StopPodSandbox() outside the kubelet will do.  As soon as the kubelet becomes aware, it will try to restart the pod, assuming the restartPolicy isn't `Never`.  So any attempt to do something while the sandbox was down would race with the kubelet bring it back up, right?

dcbw commented:
> @knobunc @dcbw i guess i'm still not complete clear on what a StopPodSandbox() outside the kubelet will do. As soon as the kubelet becomes aware, it will try to restart the pod, assuming the restartPolicy isn't Never. So any attempt to do something while the sandbox was down would race with the kubelet bring it back up, right?

@sjenning kubelet assumes everything is fine with networking on the pod at startup if kubelet/openshift-node is restarted for any reason (manual or perhaps it segfaulted/paniced).  It simply has no facility to check network status after a pod has started.  The pathological case is:

```
systemctl restart openvswitch
systemctl restart atomic-openshift-node
(or even 'yum upgrade openvswitch', which restarts OVS)
```

which blows away the OVS flows that openshift-sdn has added.  On restart, pod networking is thus broken, but kubelet does not call into the network plugin to detect that, nor does it give the plugin a chance at startup to fix things.

So we are left trying to detect that situation and somehow fix networking ourselves.  But we cannot get the necessary details out of kubelet (like the network namespace) because we aren't called from kubelet or the runtime.  So our only solution is to somehow kill the sandbox, which causes kubelet to restart that sandbox and thus networking will be reconfigured for it.

sjenning commented:
@dcbw ah I see.  so you are _counting_ on the kubelet restarting the pod.  That's actually want you want.  In that case, I don't see a problem.  lgtm.  Thanks for the explaination :)

knobunc commented:
[merge]

smarterclayton commented on pkg/sdn/plugin/runtime.go:
```diff
@@ -0,0 +1,55 @@
+	}
+
+	// Kubelet starts asynchronously and when we get an Update op, kubelet may not have created runtime endpoint.
+	// So try couple of times before bailing out (~5 second timeout).
```
You should wait longer.   Heavily contended systems could flake because of this.

smarterclayton replied:
I would expect 30s

pravisankar commented:
Updated getRuntimeService() timeout from 2 secs to 30 secs.

smarterclayton commented:
What happens to this for CRI-O?

smarterclayton commented:
Is this now fixed in the kube upstream?
